### ISSUE TITLE
Exceptions from semaphore::semaphore aren't always caught

### ISSUE THREAD

**Issue by captainurist** [kcat/openal-soft#886]
This is the stack that I'm seeing:
```
al::semaphore::semaphore(unsigned int) alsem.cpp:100
al::semaphore::semaphore(unsigned int) alsem.cpp:99
ContextBase::ContextBase(DeviceBase *) context.cpp:24
ALCcontext::ALCcontext(al::intrusive_ptr<…>, std::bitset<…>) context.cpp:128
ALCcontext::ALCcontext(al::intrusive_ptr<…>, std::bitset<…>) context.cpp:129
alcCreateContext(ALCdevice *, const ALCint *) alc.cpp:2663
// user code below
```

If `al::semaphore::semaphore` throws, then the exception goes uncaught all the way to `alcCreateContext`, which is a C function and cannot throw. On mac this ends the program with `terminate()`.

I'm running my code on MacOS on M1, not sure if the problem will be the same on other platforms, but throwing from alcCreateContext is definitely not a good idea.

**Comment by kcat** [comment]
Indeed, `alcCreateContext` shouldn't throw. It seems `sem_init` is failing when constructing the context. It should be easy to catch the exception and have the function return failure, though the device wouldn't be usable if you can't create a context since it can't create a semaphore. There's a deeper issue here that needs addressing.

Out of curiosity, what version of MacOS are you using or targeting? Ideally it should be using the `dispatch_` API since Mac's support of the `sem_` API is poor (though I don't know why it wouldn't work at all). And it should use the dispatch API if
```cpp
#if (((MAC_OS_X_VERSION_MIN_REQUIRED > 1050) && !defined(__ppc__)) || TARGET_OS_IOS || TARGET_OS_TV)
```
is true. Maybe `MAC_OS_X_VERSION_MIN_REQUIRED` isn't set high enough?

**Comment by captainurist** [comment]
Well, it's using `sem_` API because I have commented out the `#ifdef`s while experimenting with a GCC build. So my understanding is that normally this shouldn't happen. Still, I think it would make sense to catch the exception rather than terminate.

**Comment by kcat** [comment]
The crash should be fixed with commit 8b3a9b527d87009f86ed0b4eb6c33a9e8dfc069b now, but `alcCreateContext` will fail instead. Given commit a91ac01004332d59370c5d6210e1605cb8f708c6, it seems certain macOS platforms or platform versions do require using the dispatch API, as POSIX semaphores are unimplemented, so such configurations are unsupported regardless.

**Comment by captainurist** [comment]
Works now, thanks!
